rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+
+Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).
+
+Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:
+
+- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
+- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
+- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
+
+## Proposal
+
+GOV.UK will adopt a container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
+
+GOV.UK applications will be built into container images, which will include the operating system. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
+
+A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
```

Comment:
Is this referring to upgrading the host OS?

Reply:
Given that the "where" these contained processes are going to run is out of scope for the RFC, we could potentially end up not managing the operating system on which they run, so I'm also unsure what this refers to by operating system version.

I do see a reduction in risk though if you can more tightly manage the applications dependencies, rather than knowing it's probably dependent on _some_ Ubuntu packages or bits of Ubuntu.

Reply:
Yeah, this paragraph is based on the (unchecked) assumption that the container would include an operating system as a dependency (a la `FROM ubuntu` or using the `cflinuxfs3` stack that GOV.UK PaaS provides), which would then be easier to upgrade (than the existing infrastructure which has multiple applications sharing the same OS).

The idea is very much that we wouldn't end up managing the host operating system on which the containers run, leaving that to whichever platform we end up with.

I guess we should reword this along the lines of:

```suggestion
The most immediate benefit of this approach is that it will allow us to offload upgrades of the host operating system to whomever manages the container runtime, and only have to manage updates to the applications' dependencies within the containers. By using containers to manage application dependencies more tightly, GOV.UK can enable a much lower risk upgrade path. 
```

Reply:
Personally, I would see this in terms of separation between the applications and the operating system, and having the opportunity to be more specific about what applications require and use. I'm not sure how related this RFC is to offloading responsibility for managing stuff at the operating system level.

Reply:
> The idea is very much that we wouldn't end up managing the host operating system on which the containers run, leaving that to whichever platform we end up with.

I didn't get this from reading the proposal section. In fact the wording suggests otherwise "
GOV.UK will adopt a container based hosting environment", in that GOV.UK will have responsibility for the hosting environment, rather than just "begin using a container based hosting environment".

Reply:
I've suggested a change in wording in the proposal section to clear that up.

I think we'd like to avoid having responsibility for the hosting environment if at all possible.

Reply:
> I've suggested a change in wording in the proposal section to clear that up.
> 
> I think we'd like to avoid having responsibility for the hosting environment if at all possible.

Cool, I see that the proposal is now to "use a managed" platform.

Returning to the discussion of what "incrementally upgrade the operating system version" means in the context of containers, I like the change you proposed Richard. Hopefully if we start working at the "container" (which I read as process) level, and doing this on a managed platform, then there wouldn't be any operating systems to directly interact with.

Reply:
> there wouldn't be any operating systems to directly interact with.

I guess it's a definitional thing. If we're running ruby apps (and let's definitely assume we are), we're going to need at least need glibc (or musl), openssl etc., which are "Operating system" level things.

We can think of these things as dependencies (rather than bits of an OS) if we like, but we'll still have to think about monitoring them for vulnerabilities and patching them and so on.

Given that, we _might_ decide to put an OS distribution in the container, because then we can piggy back on the distribution's patch schedule. Or we might not, because the fewer things we depend on the fewer things we need to patch. Or we might decide to use the cloud foundry approach, and build "droplets" instead of full images, in which case the operating system really doesn't live in the image, but gets added to the container at runtime. All interesting options with their own pros and cons.

I think that decision is out of scope for this RFC though - suffice to say we'll still have to patch things inside the containers, but that should be easier than doing an upgrade of an OS on a machine with multiple apps running on it.

Reply:
I've committed the suggestion above in 9aa241450f1e0365443e30b9e2eb151098c79586, hopefully that clarifies this point enough.

Marking this conversation as "Resolved" to try and clear up the RFC, but let me know if it needs more clarification :)